### PR TITLE
Fix friends crash

### PR DESCRIPTION
**Pull request by nada-ezziti**
#### **Description**:
This pull request addresses an issue where parts of the code defaulted to `"unknown"` when a user was not logged in. This behavior caused an `"unknown"` user to be created in the database, leading to crashes when attempting to add the `"unknown"` account as a friend. The following changes have been made:

1. **Removed all uses of the `"unknown"` default value**:
   - Operations now fail instead of defaulting to `"unknown"`.
   - Proper error handling is implemented to handle missing or invalid user data.

2. **Ensured invalid `userId`s are handled safely**:
   - Introduced checks for null, blank, or invalid `userId`s.
   - Improved error logging and flow to prevent crashes.

#### **Files Changed**:
- **`FirestoreUserRepository.kt`**: Updated `getUserName` to remove `"unknown"` fallback and handle errors properly.
- **`MockUserRepository.kt`**: Adjusted mock logic to align with the updated behavior of failing for invalid or missing users.
- **`SaveUserToFirestore.kt`**: Removed the `"unknown"` fallback for user email or name, added proper validation, and ensured no invalid users are saved to Firestore.
- **`UserViewModel.kt`**: Updated to align with the new handling of user data and avoid crashes.

3. **Test Updates**:
- **`SaveUserToFirestoreTest.kt`**:  
  - Added tests to verify proper logging and behavior when `email` is `null` or blank.  
  - Ensured Firestore operations are skipped when user validation fails.

- **`FirestoreUserRepositoryTest.kt`**:  
  - Updated tests to validate that errors are handled correctly and `"unknown"` is no longer used.

- **`MockUserRepositoryTest.kt`**:  
  - Adjusted existing test cases to reflect the removal of `"unknown"`.  
  - Added edge case tests for invalid or missing user data.

#### **Issue Fixed**:
- Fixes the bug where the app crashes when attempting to add an `"unknown"` user as a friend.

### CODE REVIEW

File: app/src/main/java/com/github/se/signify/model/common/user/MockUserRepository.kt [hunk 1]
```diff
@@ -92,11 +92,15 @@ class MockUserRepository : UserRepository {
...
-    onSuccess(user.name ?: "unknown")
+    val user = users[userId]
+    if (user != null) {
+      // Ensure user.name is non-nullable before passing it to onSuccess
+      user.name?.let { onSuccess(it) }
+          ?: onFailure(NoSuchElementException("Username is missing for user ID: $userId"))
+    } else {
+      // Handle the case where the user is not found
+      onFailure(NoSuchElementException("User not found for ID: $userId"))
+    }
```
Comment (samymessouak) [inline app/src/main/java/com/github/se/signify/model/common/user/MockUserRepository.kt:103]: Maybe it is possible to combine the if-else in a single block for a better flow. 
Watch out if this code handles correctly the null verification, else it is also good the way you did it.

```suggestion
   val user = users[userId] ?: return onFailure(NoSuchElementException("User not found for ID: $userId"))
user.name?.let { onSuccess(it) }
    ?: onFailure(NoSuchElementException("Username is missing for user ID: $userId"))

```

File: app/src/test/java/com/github/se/signify/model/common/user/FirestoreUserRepositoryTest.kt [hunk 2]
```diff
@@ -1377,4 +1377,114 @@ class FirestoreUserRepositoryTest {
...
+      failureCallbackCalled = true
+      assertTrue(exception is NoSuchElementException)
+      assertEquals(expectedException.message, exception.message)
+    }
+
+    // Act
+    firestoreUserRepository.getUserName(
+        currentUserId,
+        onSuccess = {
+          fail("onSuccess should not be called when the username is missing or empty")
```
Comment (samymessouak) [inline app/src/test/java/com/github/se/signify/model/common/user/FirestoreUserRepositoryTest.kt:1411]: You are using this message twice, maybe you can define it in a constant.

File: app/src/test/java/com/github/se/signify/model/common/user/SaveUserToFirestoreTest.kt [hunk 3]
```diff
@@ -125,4 +130,76 @@ class SaveUserToFirestoreTest {
...
+      // Prevent Firestore from initializing
+      val mockFirestore = mock(FirebaseFirestore::class.java)
+      `when`(FirebaseFirestore.getInstance()).thenReturn(mockFirestore)
+
+      // Act
+      saveUserToFirestore()
+
+      // Verify Log.e is called with the correct message
+      logMock.verify(
+          { Log.e(eq("FireStore"), eq("Cannot save user: email is null or blank")) }, times(1))
```
Comment (samymessouak) [inline app/src/test/java/com/github/se/signify/model/common/user/SaveUserToFirestoreTest.kt:194]: You are also using the string "Cannot save user: email is null or blank" twice, you can also define it in a constant.

**Review by samymessouak — APPROVED** [review]
## Strengths to Highlight
**1.Removal of "unknown" Default Handling**

- The PR resolves a critical bug where "unknown" user data caused crashes.

- Good approach in failing explicitly (NoSuchElementException) instead of silently defaulting to invalid values.

**2. Improved User Validation**

- Validation for null/empty userId, email, and user.name makes the code safer and reduces unexpected edge cases.
- Logging errors (Log.e) when email or userId is invalid is a good practice for debugging.

**3. Comprehensive Test Cases**

Tests are well-structured and cover all edge cases:
-  Null, blank, and valid user data.
-  Proper handling when a user does not exist.


**4. Readability and Error Messaging**

- Error messages like "User not found for ID: $userId" and "Username is missing or empty for ID: $userId" are clear and informative.

**I requested some change in comments but I dont think it is that important as your PR is already really good. I approved it and feel free to watch what i suggested !**
